### ISSUE TITLE
Builds without OpenSSL

### ISSUE THREAD

**Issue by vks** [timvisee/ffsend-api#2]
It would be nice to have to build ffsend using ring instead of OpenSSL. This would likely simplify the build on Windows and use a more minimal crypto library.

As far as I can see, this requires the following changes:

* Make `reqwest` use `rustls`:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index ad5df53..f6259c1 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -48,7 +48,7 @@ mime = "0.3.12" # same as reqwest
 mime_guess = "2.0.0-alpha.4"
 openssl = "0.10"
 regex = "1.1"
-reqwest = "0.9.2"
+reqwest = { version = "0.9.2", features = ["rustls"] }
 serde = "1.0"
 serde_derive = "1.0"
 serde_json = "1.0"
```
* Port random data generation to ring.
* Port signing to ring.
* Port encryption to ring.

The last part is likely the most tedious.

**Comment by timvisee** [comment]
I've tried this before but canceled it because it was too much work.

I'm in for it though, implementing this would be awesome! I think the requirements you mentioned are spot on. The crypto logic is currently isolated quite well in the [src/crypto](./src/crypto) directory.

Currently I want to focus on other things first, and might start with this at a later time. Working on a `ring` branch along with master to slowly implement `ring`support would probably be a good idea.

If there's someone else that wants to start with this, go for it!

**Comment by timvisee** [comment]
I've created the following issue for this: https://gitlab.com/timvisee/ffsend-api/issues/6

Work will be made in the [`crypto-ring`](https://gitlab.com/timvisee/ffsend-api/tree/master) branch.

**Comment by vks** [comment]
Thanks for the follow up! I ported some parts to ring on my branch, you might want to use that:

* Symmetric encryption: https://github.com/vks/ffsend-api/commit/a39296c708c7add1dd0ffe3ad578bc9b4926a126
* Signing: https://github.com/vks/ffsend-api/commit/d9da694e27eef1855fa8aa6267129a0c0b587b44

**Comment by timvisee** [comment]
Awesome, thanks for your changes! This looks perfect.

I would like to note that the primary repository is on GitLab infact, this repository on GIthub is just a mirror (the `crypto-ring` branch is only on GitLab). I'll probably merge your changes over there.

I'll be sure to take a look into this later this day after work.

**Comment by timvisee** [comment]
I've implemented your changes in [`crypto-ring`](https://gitlab.com/timvisee/ffsend-api/commits/crypto-ring) with conditional compilation (to support both back-ends).

Now, what's left is some cryptography using `openssl` in `src/pipe/crypto/`:

![image](https://user-images.githubusercontent.com/856222/54445332-d3e72580-4744-11e9-8a78-449b31252594.png)

**Comment by timvisee** [comment]
This has now been implemented. Ring is now used as default crypto backend.

This is released as part of `v0.6.0`.

See the issue over on GitLab for more details: https://gitlab.com/timvisee/ffsend-api/-/issues/6
